cli/init.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package cli
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+func MakeInitCommand() *cobra.Command {
+	var encrypted bool
+	var cmd = &cobra.Command{
+		Use:   "init",
+		Short: "Create root directory and configuration file",
+		Long: `Create root directory and configuration file
+
+Example: with encryption at rest
+  defradb init -e
+		`,
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+			return setContextRootDir(cmd)
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			rootdir := mustGetContextRootDir(cmd)
+			if err := os.MkdirAll(rootdir, 0755); err != nil {
+				return err
+			}
+
+			cfg := defaultConfig()
+			cfg.AddConfigPath(rootdir)
+
+			err := bindConfigFlags(cfg, cmd.Root().PersistentFlags())
+			if err != nil {
+				return err
+			}
+
+			if encrypted {
+				// generate a random encryption key
+				key := make([]byte, 32)
```

Comment:
suggestion:  I think it is worth documenting the `32`, as if it is too small it could be insecure no?

thought: Is it worth us taking ownership over key generation like we are here?  Might it not be better to just let users provide their own via a CLI arg?

Reply:
> thought: Is it worth us taking ownership over key generation like we are here? Might it not be better to just let users provide their own via a CLI arg?

This is just for convenience. It is possible to run `defradb init` and then provide your own encryption key in the config.

Reply:
Oh, I missed something - we are just storing the key in plain text in the config... Are we sure we want to do that even in the short term?

Reply:
I think in the future we could use Orbis for key management. I don't see an issue with storing it in the config. My thought process being: if you can access the config you can access the data files as well.

Reply:
The issue I see with keeping it in the config is that it might make it more difficult for projects that like to store config in their repos. This would could lead to keys being shared with people that shouldn't have access to them.

Reply:
>  if you can access the config you can access the data files as well

The point of this PR is that the data is encrypted on disk, so if someone gains access to the machine they cant access the data no?  And we are storing the key in plain text next to the 'protected' data.

Reply:
There is also the automatic environment variable for that case.

Reply:
> There is also the automatic environment variable for that case.

That has similar problems to storing it in the config, if someone has access to the file system, they presumably have access to the env vars too?

Reply:
> That has similar problems to storing it in the config, if someone has access to the file system, they presumably have access to the env vars too?

I think the key should be in a separate file to save the potential config headache (key in repo for as stated above). That will have the same issue as storing it in the `env` variables though. One that we should not think we are responsible for. The level of safety needed depends on the deployment environment and the developers should be the one responsible for their usage of secret keys. All that we need to worry about is providing options for them to pass their keys along to Defra.

Reply:
Discussed in the discord thread, but small note for @AndrewSisley

> suggestion: I think it is worth documenting the 32, as if it is too small it could be insecure no?

32bytes is the correct key size for AES256. It ideally runs with 128, 192, or 256 *bits*, which 32bytes*8 = 256 bits

Reply:
> 32bytes is the correct key size for AES256. It ideally runs with 128, 192, or 256 bits, which 32bytes*8 = 256 bits

Yes, I trusted the current number was appropriate, I'm more worried about documenting it so that nobody tries to set it to something smaller.